### ISSUE TITLE
Add ability to parallel copy with per-host remote & local file paths

### ISSUE THREAD

**Issue by Shados** [ParallelSSH/parallel-ssh#110]
### Motivation
I'd like the the ability to specify per-host remote and local file paths for `ParallelSSH2Client`'s `copy_remote_file` function (or possibly as a separate function). My specific use-cases:

#### Per-host remote paths
I have a Linux script which generates a file with a randomized path (specifically, using `mktemp`) and outputs the path on stdout. 

Right now, it's easy enough for me to run said script in parallel across my list of hosts and capture the output paths from each, but I can't easily copy each file produced back to my local host in parallel.

#### Per-host local paths
I have a set of shared files I need to retrieve from a list of hosts, but the local paths I want to store them in differ non-trivially per-host (the list of hosts is logically divided into various groups and sub-groups, and I want to store the files into a somewhat complicated directory tree reflecting this).

Obviously, I can always just retrieve the files to a single location with the current code and move them from there to their final local paths, but this presents at least two downsides:
1. I'd essentially have to replicate existing work to check for files which failed to copy + handle the file moves in parallel, making the code much less clean than if parallel-ssh were simply copying them to the correct path to begin with.
2. While not an issue for me right now, if I were using this to move large files around I could potentially fill up the fs for my single local directory. Whereas, directly copying to multiple local paths would allow me to spread the space requirement across several mounts instead of needing a single mount with enough space to temporarily hold all the files.

------

### Implementation
I've taken a look at the existing code and I'd be happy enough to implement this functionality myself, but I'm not sure what kind of interface would be preferred for it. I can see two obvious approaches:

1. Add a `host_paths` optional argument to `copy_remote_function`, taking something similar to `host_config` (see below), with un-specified hosts/paths in it defaulting to the normal `remote_file` and `local_file` argument behaviour.
```
host_paths = {
    'host1': {'remote_file': 'a_file',
              'local_file':  'host1/a_file',
              'recurse':      True},
    'host2': {'remote_file': 'b_file'},
    'host3': {'local_file':  '/another/dir/host3_file'},
}
```
2. Add a separate function entirely to handle parallel per-host copies, and either directly take something like the above `host_paths` structure as an argument (either ignoring or (optionally?) erroring for un-specified hosts), or possibly take a callback function run per-host to generate the paths/options

**Comment by pkittenis** [comment]
Hi there,

Thanks for the interest and report. It sounds like a useful feature and have also been thinking of ways to improve the current parallel copy filename functionality.

First, let's see how this can be done currently, with a bit more code to write:

```
from gevent import spawn, joinall

client = ParallelSSHClient(<..>)

host_paths = {
    'host1': {'remote_file': 'a_file',
              'local_file':  'host1/a_file',
              'recurse': True},
    'host2': {'remote_file': 'b_file', 'local_file': 'host2_b_file'},
    'host3': {'local_file':  '/another/dir/host3_file'},
}

copies = [spawn(client.host_clients[host].copy_remote_file, 
          host_paths[host]['remote_file'], 
          host_paths[host]['local_file'], 
          recurse=host_paths[host].get('recurse', False})
    for host in host_paths]
joinall(copies, raise_error=True)
```

`client.host_clients[host]` is the `SSHClient` for that host giving direct access to its `copy_remote_file` function and allows for both filenames to be specified. 

This pattern can be used to extend parallel client with new functions, overloading existing ones and so on.

That should give a workaround while direct support is implemented.

As for implementation, here is what I had in mind for this functionality:

* Allow a string pattern to be specified for either or both local and remote file with formatting arguments to be used, like `host_args` command substitution functionality. This can include sub directories. Similar to the above proposal but with formatting for string substitution rather than overrides.

Example use:

```
copy_args = [
    {'remote_file': 'file_a', 'local_file': 'a_sub_dir/host1_file'}
    {'remote_file': 'file_b', 'local_file': 'file_b'}
]

client.copy_remote_file('%(remote_file)s', '%(local_file)s', 
    recurse=True, copy_args=copy_args)
```

Where `client.hosts[0]` would get remote and local file format arguments from `copy_args[0]` and so on. Multiple formatting arguments can also be used. Must have same number of `copy_args` as number of hosts.

How does this look for the use case? Would prefer a substitution method rather than optionally overriding as it is more explicit with what is supposed to happen when there is a string substitution format in the file arguments.

Otherwise, a new function only taking per-host arguments would be best IMO.

**Comment by Shados** [comment]
Yep, I'd already implemented a similar interim work-around, except by monkey-patching in a separate function for it, as it's getting called from a few different places in my code.

Your example using string substitution looks fine to me, and keeping consistent with your existing similar functionality in `run_command` is obviously desirable.

**Comment by Shados** [comment]
Went to implement this today, but ran into some issues with running the tests under Nix (the package manager / build system I am using) and got distracted by investigating and opening an issue there. I'll run the tests in a more typical environment on Monday and submit a PR.
